src/Castle.Core/DynamicProxy/Generators/MetaMethod.cs

```diff
@@ -101,7 +102,30 @@ internal override void SwitchToExplicitImplementation()
 				Attributes |= MethodAttributes.SpecialName;
 			}
 
-			name = string.Format("{0}.{1}", Method.DeclaringType.Name, Method.Name);
+			var nameBuilder = new StringBuilder();
+			AppendTypeNameTo(nameBuilder, Method.DeclaringType);
+			nameBuilder.Append('.');
+			nameBuilder.Append(Method.Name);
+			name = nameBuilder.ToString();
+		}
+
+		private static void AppendTypeNameTo(StringBuilder stringBuilder, Type type)
+		{
+			stringBuilder.Append(type.Name);
+			if (type.GetTypeInfo().IsGenericType)
+			{
+				stringBuilder.Append('[');
+				var genericTypeArguments = type.GetGenericArguments();
+				for (int i = 0, n = genericTypeArguments.Length; i < n; ++i)
```

Comment:
Was just basing that on a rudimentary find of usages for `MetaMethod`. 

![image](https://user-images.githubusercontent.com/607402/27842605-45601a5c-6103-11e7-837b-64dc7da56ba3.png)